### ISSUE TITLE
Add Anti-DDoS module

### ISSUE THREAD

**Issue by x0rzkov** [bunkerity/bunkerized-nginx#2]
HI,

Hope you are all well !

Is it possible to add an Anti-DDoS module to Nginx ?

Found that: https://github.com/C0nw0nk/Nginx-Lua-Anti-DDoS

What do you think ? Would it impact the SEO ?

Cheers,
X

**Comment by bunkerity** [comment]
Looks like a good idea. Maybe we could add a reverse DNS checker to whitelist search engines crawlers. Give me some time to implement and test this.

**Comment by DistractionRectangle** [comment]
From the looks of it, that won't protect you from DDoS attacks. The idea seems to be generate a computationally cheap puzzle to check if the visitor is actually rendering the page (running the js and solving the challenge). If you're being flooded (level 4), this doesn't stop them from flooding you and it doesn't waste their resources. This may slow application level attacks (level 7) down (barely), but it appears to be mostly an exercise rather than something that's actually deployed.

Further, it doesn't have releases and implements a lot of functionality offered by other packages in this repo from scratch. Including it would mean either maintaining two sets of the same configuration or manually disabling 90%+ of that module as well as stability testing to figure out if it works properly//doesn't conflict.

As far as SEO goes, enabling it will definitely impact SEO unless you manually disable it for search engine crawlers. 

In short, it won't work on for level 4 DDoS attacks, and barely does anything for level 7 DDoS attacks. Dialing in Fail2ban will be effective at mitigating level 7 DDoS, and the only real solution for level 4 is having more bandwidth or control over routing so it doesn't hit your server. It might be okay at other things, but that functionality is better left to upstream projects dedicated to doing those things specifically. Cloudflare + fail2ban will get you 99.9+% of the way there, and for that last bit, just toggle Cloudflare's Under Attack Mode.

**Comment by x0rzkov** [comment]
Is there any known alternative on github ?

**Comment by DistractionRectangle** [comment]
Regarding layer 7, sure. Fail2ban, WAFs, captchas etc work at the application level. Fail2ban and ModSecurity are both well documented and actively maintained.

For simple DoS attacks that merely overwhelm your bandwidth, there's no great software layer solution, you need more bandwidth and/or control over your routes. You can build out for own infrastructure, setup auto scaling, etc but that's a lot of hats, not to mention cost, for most individuals//small businesses. For a peek into how enterprises deal with this, look of DDoS postmortems, e.g. https://github.blog/2018-03-01-ddos-incident-report/

**Comment by bunkerity** [comment]
As stated by @DistractionRectangle it won't be usefull as an anti-ddos solution for sure. But it can be used to block automated bots/scanners and make the task harder for script kiddies to exploit a public vulnerability. IMO, adding something like this into bunkerized-nginx won't hurt especialy if we can easily turn it off and manage it through environment variable. I didn't see any ready to use project so some extra work is needed.

**Comment by bunkerity** [comment]
I've added a new feature called "antibot" to force users doing a "challenge" before they can access the web service. There are, at the moment, 4 differents kinds of challenges :
- cookie : basic Set-Cookie/Cookie test
- javascript : basic javascript challenge
- captcha : text captcha (no external dependency)
- recaptcha : use reCAPTCHA v3 from Google and deny/allow if score is ok

More info [here](https://github.com/bunkerity/bunkerized-nginx#antibot).

Feel free to open a new issue if you have any questions or suggestions.
